src/Compilers/CSharp/Portable/Lowering/InitializerRewriter.cs

```diff
@@ -21,56 +21,81 @@ internal static BoundTypeOrInstanceInitializers RewriteConstructor(ImmutableArra
+                        (object)expression.Type != null &&
+                        expression.Type.SpecialType != SpecialType.System_Void)
                     {
-                        submissionResult = expr;
+                        trailingExpression = expression;
                         continue;
                     }
                 }
+
                 boundStatements.Add(RewriteInitializersAsStatements(initializer));
             }
 
-            var syntax = method.GetNonNullSyntaxNode();
-            if ((object)submissionResultType != null)
+            if (hasSubmissionResultType && (trailingExpression != null))
             {
-                if (submissionResult == null)
+                Debug.Assert(submissionResultType.SpecialType != SpecialType.System_Void);
+
+                // Note: The trailing expression was already converted to the submission result type in Binder.BindGlobalStatement.
+                if (labeledStatement == null)
                 {
-                    // Return default(T) if submission does not have a trailing expression.
-                    submissionResult = new BoundDefaultOperator(syntax, submissionResultType);
+                    statement = new BoundReturnStatement(statement.Syntax, trailingExpression);
                 }
-                Debug.Assert(submissionResult.Type.SpecialType != SpecialType.System_Void);
-
-                // The expression is converted to the submission result type when the initializer is bound.
-                boundStatements.Add(new BoundReturnStatement(submissionResult.Syntax, submissionResult));
+                else
+                {
+                    statement = new BoundReturnStatement(labeledStatement.Body.Syntax, trailingExpression);
+                    statement = new BoundLabeledStatement(labeledStatement.Syntax, labeledStatement.Label, statement);
+                }
+                boundStatements.Add(statement);
+                hasTrailingExpression = true;
             }
-
-            return new BoundTypeOrInstanceInitializers(syntax, boundStatements.ToImmutableAndFree());
+            else
+            {
+                hasTrailingExpression = false;
+            }
+    
+            return new BoundTypeOrInstanceInitializers(method.GetNonNullSyntaxNode(), boundStatements.ToImmutableAndFree());
         }
 
         /// <summary>
-        /// Returns the expression if the statement is actually an
-        /// expression (ExpressionStatementSyntax with no trailing semicolon).
+        /// Returns the expression if the statement is actually an expression (ExpressionStatementSyntax with no trailing semicolon).
         /// </summary>
-        internal static BoundExpression GetTrailingScriptExpression(BoundStatement statement)
+        internal static BoundExpression GetTrailingScriptExpression(BoundStatement statement, out BoundLabeledStatement labeledStatement)
         {
+            if (statement.Kind == BoundKind.LabeledStatement)
```

Comment:
Should this be ```while``` rather than ```if```? Instead, we should consider reporting a parse error if a LabeledStatement contains an expression.